extensions/notebook/src/book/bookTreeView.ts

```diff
@@ -785,5 +801,52 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		return localRoots;
 	}
 
+	public async undo(): Promise<void> {
+		if (this.undoTocFiles?.length > 0 && this.undoMovedFiles?.length > 0) {
+			let files = this.undoMovedFiles.pop();
+			this.redoMovedFiles.push(files);
+			this.redoTocFiles.push(this.undoTocFiles.pop());
+			let tocFiles: Map<string, JupyterBookSection[]> = this.undoTocFiles.pop();
+			// restore toc files
+			for (const [tocPath, contents] of tocFiles.entries()) {
+				await this.applyTocChanges(tocPath, contents);
+			}
+			// return files to previous file path
+			for (const [src, dest] of files.entries()) {
```

Comment:
So if I move a file, and then move it again what's going to happen? Because these aren't stored in time order and so you can't guarantee which move operation will come "first" which could result in trying to undo the move of a file that doesn't exist (yet)